### PR TITLE
Sync of upstream changes from substrate-node-template

### LINKED ISSUE #63: Quarterly Uptake of new upstream changes from substrate-node-template
We have to track and make sure we are compatible with latest updates to Node templates.

Do Quarterly uptake of upstream main brain into our own universal-develop branch.

https://docs.github.com/en/pull-requests/collaborating-with-pull-requests/working-with-forks/syncing-a-fork#syncing-a-fork-branch-from-the-command-line

### PR DESCRIPTION
**Pull request by evilrobot-01**
Synchronised fork with upstream, updating pallets to resolve issues resulting from breaking changes. 

A few notes/todos:

- [ ] `#[pallet::unbounded]` used on all storage items which used unbounded vectors. This was to reduce the complexity of the update to latest and another issue should be created to tackle.
- [X] Used `bounded_*()` helper functions for Dao tests, but only as a wrapper until the pallet can be fully migrated to bounded vecs.
- [ ] assess public visibility of helper functions in `dao` pallet
- [ ] ensure that usage of `into_account_truncating()` is correct within `task` pallet
- [ ] set max lengths of new config items of `dao` config in runtime. These are used by the Dao struct and were previously unbounded.
- [ ] there were upstream changes to RPC apis (`jsonrpsee`) , need to check that all still works from the front-end. 
- [ ] review possible switch to new `list_benchmarks!` and `add_benchmarks!` macros within runtime benchmarks
- [ ] Cargo.lock should not be included within source control for library crates (pallets) : [source](https://github.com/rust-lang/cargo/issues/315#issuecomment-634598379)

Commits below show upstream first and then my merge and pallet update commits.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -35,7 +35,7 @@ First, complete the [basic Rust setup instructions](./docs/rust-setup.md).
 Use Rust's native `cargo` command to build and launch the template node:
 
 ```sh
-cargo run --release -- --dev --tmp
```
Comment (stojanov-igor) [inline README.md:38]: why --tmp is removed?

File: pallets/dao/Cargo.toml [hunk 2]
```diff
@@ -4,104 +4,105 @@ version = '0.7.0'
...
-
-[dependencies.sp-core]
-default-features = false
-git = 'https://github.com/paritytech/substrate.git'
-tag = 'monthly-2021-11'
-version = '4.0.0-dev'
+version = '3.0.0'
 
 [dependencies.frame-benchmarking]
 default-features = false
```
Comment (stojanov-igor) [inline pallets/dao/Cargo.toml]: Should we use consisted inline format everywhere just like you have done in node/Cargo.toml?

File: pallets/dao/src/tests.rs [hunk 3]
```diff
@@ -11,22 +11,47 @@ fn vision() -> Vec<u8> {
 	vec![1u8, 7]
 }
 
+fn bounded_vision() -> BoundedVec<u8, MaxVisionLen>
```
Comment (stojanov-igor) [inline pallets/dao/src/tests.rs:14]: This is better

File: pallets/did/Cargo.toml [hunk 4]
```diff
@@ -4,30 +4,89 @@ version = '0.7.0'
...
-edition = '2018'
+edition = '2021'
 license = 'Apache-2.0'
 publish = false
 repository = "https://github.com/UniversalDot/universal-dot-node"
 
 [package.metadata.docs.rs]
 targets = ["x86_64-unknown-linux-gnu"]
 
-[dependencies]
```
Comment (stojanov-igor) [inline pallets/did/Cargo.toml:15]: Inline does look cleaner

File: pallets/profile/src/lib.rs [hunk 5]
```diff
@@ -81,8 +81,8 @@ pub mod weights;
 #[frame_support::pallet]
 pub mod pallet {
 	use frame_support::{dispatch::DispatchResult,
-	storage::bounded_vec::BoundedVec,
-	pallet_prelude::*};
+						storage::bounded_vec::BoundedVec,
```
Comment (stojanov-igor) [inline pallets/profile/src/lib.rs]: What is happening with the spacing here?

File: pallets/task/src/lib.rs [hunk 6]
```diff
@@ -298,7 +298,7 @@ pub mod pallet {
 			// Transfer balance amount to escrow account
 			let sub_account = Self::account_id(&task_id);
 			<T as self::Config>::Currency::transfer(&signer, &sub_account, budget,
-				ExistenceRequirement::KeepAlive)?;
+													ExistenceRequirement::KeepAlive)?;
```
Comment (stojanov-igor) [inline pallets/task/src/lib.rs:301]: THe spacing has to be fixed here.. your code editor is spacing things automatically I guess

File: pallets/task/src/lib.rs [hunk 7]
```diff
@@ -325,11 +325,11 @@ pub mod pallet {
 				if task.budget > budget {
 					let difference = task.budget - budget;
 					<T as self::Config>::Currency::transfer(&sub_account, &signer, difference,
-						ExistenceRequirement::KeepAlive)?;
+															ExistenceRequirement::KeepAlive)?;
```
Comment (stojanov-igor) [inline pallets/task/src/lib.rs]: Also spacing here

File: runtime/Cargo.toml [hunk 8]
```diff
@@ -1,189 +1,54 @@
...
-tag = 'monthly-2021-11'
-version = '4.0.0-dev'
-
+targets = ["x86_64-unknown-linux-gnu"]
+
+[dependencies]
+codec = { package = "parity-scale-codec", version = "3.0.0", default-features = false, features = ["derive"] }
+scale-info = { version = "2.1.1", default-features = false, features = ["derive"] }
+
+pallet-aura = { version = "4.0.0-dev", default-features = false, git = "https://github.com/paritytech/substrate.git", branch = "polkadot-v0.9.27" }
```
Comment (stojanov-igor) [inline runtime/Cargo.toml:19]: Inline is better. We should use it in our pallets also.

**Review by stojanov-igor — CHANGES_REQUESTED** [review]
My Build fails with: 

`error: failed to select a version for the requirement `libp2p = "^0.46.1"`
candidate versions found which didn't match: 0.45.1, 0.45.0, 0.44.0, ...
location searched: crates.io index
required by package `sc-cli v0.10.0-dev (https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)`
    ... which satisfies git dependency `sc-cli` (locked to 0.10.0-dev) of package `node-template v4.0.0-dev (/Users/igorstojanov/code/github-universal-dot/universal-dot-node/node)``

We can use the inline formaing for cargo packages.. Also minor fixes for spacing.

File: README.md (hunk 1 above)
Comment (evilrobot-01) [reply README.md:38]: Upstream - https://github.com/substrate-developer-hub/substrate-node-template/commit/0fa941470d70c03169cedb08213fca6247c141f9

File: pallets/profile/src/lib.rs (hunk 5 above)
Comment (evilrobot-01) [reply pallets/profile/src/lib.rs]: That's why I normally just use Rustfmt on save. Tell me what settings you use for formatting and I will replicate. Manually re-formatting now.

File: pallets/task/src/lib.rs (hunk 6 above)
Comment (evilrobot-01) [reply pallets/task/src/lib.rs:301]: Restored in next commit

File: pallets/task/src/lib.rs (hunk 7 above)
Comment (evilrobot-01) [reply pallets/task/src/lib.rs]: Resolved in next commit.

File: pallets/profile/src/lib.rs (hunk 5 above)
Comment (evilrobot-01) [reply pallets/profile/src/lib.rs]: Resolved in next commit.

**Comment by evilrobot-01** [comment]
> My Build fails with:
> 
> `error: failed to select a version for the requirement `libp2p = "^0.46.1"`candidate versions found which didn't match: 0.45.1, 0.45.0, 0.44.0, ... location searched: crates.io index required by package`sc-cli v0.10.0-dev ([https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)`](https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)%60) ... which satisfies git dependency `sc-cli` (locked to 0.10.0-dev) of package `node-template v4.0.0-dev (/Users/igorstojanov/code/github-universal-dot/universal-dot-node/node)``
> 
> We can use the inline formaing for cargo packages.. Also minor fixes for spacing.

If you you try a `cargo clean` first?

File: pallets/dao/Cargo.toml (hunk 2 above)
Comment (evilrobot-01) [reply pallets/dao/Cargo.toml]: Switched to inline dependency style for pallets.

File: runtime/Cargo.toml (hunk 8 above)
Comment (evilrobot-01) [reply runtime/Cargo.toml:19]: Switched to inline dependency style for pallets.

**Comment by stojanov-igor** [comment]
> > My Build fails with:
> > `error: failed to select a version for the requirement `libp2p = "^0.46.1"`candidate versions found which didn't match: 0.45.1, 0.45.0, 0.44.0, ... location searched: crates.io index required by package`sc-cli v0.10.0-dev ([https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)`](https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)%60) ... which satisfies git dependency `sc-cli`(locked to 0.10.0-dev) of package`node-template v4.0.0-dev (/Users/igorstojanov/code/github-universal-dot/universal-dot-node/node)``
> > We can use the inline formaing for cargo packages.. Also minor fixes for spacing.
> 
> If you you try a `cargo clean` first?

Still having the same issue..

`error: failed to select a version for the requirement `libp2p = "^0.46.1"`
candidate versions found which didn't match: 0.45.1, 0.45.0, 0.44.0, ...
location searched: crates.io index
required by package `sc-cli v0.10.0-dev (https://github.com/paritytech/substrate.git?branch=polkadot-v0.9.27#8eff668a)`
    ... which satisfies git dependency `sc-cli` (locked to 0.10.0-dev) of package `node-template v4.0.0-dev (/Users/igorstojanov/code/github-universal-dot/universal-dot-node/node)``

**Comment by stojanov-igor** [comment]
Had an older rust version which was causing this issue.. 

Upgraded and fixed the problem. 

This can be merged.

**Comment by evilrobot-01** [comment]
> Had an older rust version which was causing this issue..
> 
> Upgraded and fixed the problem.
> 
> This can be merged.

Great to hear! I did see that was a potential issue whilst looking into the issue so glad you got it sorted. 

Any feedback on the [outstanding tasks](https://github.com/UniversalDot/universal-dot-node/pull/64#issue-1337302475) before merging? Shall I just create separate issues for each of them?

**Comment by evilrobot-01** [comment]
Outstanding tasks moved to https://github.com/UniversalDot/universal-dot-node/issues/67

**Pull request merged**